File: packages/typescript/tests/contracts/api-contract.test.ts
import { describe, it, expect, beforeAll, afterAll, afterEach } from 'vitest';
import { http, HttpResponse } from 'msw';
import { LunaClient } from '../../src/client';
import { server, resetHandlers } from '../mocks/server';
import { mockUser, mockProject, mockBucket } from '../mocks/fixtures';

const BASE_URL = 'https://api.eclipse.dev';

/**
 * Contract validation tests ensure API responses match expected schemas.
 * These tests verify the structure and types of API responses.
 */
describe('API Contract Validation', () => {
    let client: LunaClient;

    beforeAll(() => {
        client = new LunaClient({ apiKey: 'lk_test_12345678901234567890123456789012' });
    });

    describe('User contract', () => {
        it('should have required user fields', async () => {
            const user = await client.users.get(mockUser.id);

            // Required fields
            expect(user).toHaveProperty('id');
            expect(user).toHaveProperty('name');
            expect(user).toHaveProperty('email');
            expect(user).toHaveProperty('created_at');
            expect(user).toHaveProperty('updated_at');

            // Type checks
            expect(typeof user.id).toBe('string');
            expect(typeof user.name).toBe('string');
            expect(typeof user.email).toBe('string');
        });

        it('should have correct ID prefix', async () => {
            const user = await client.users.get(mockUser.id);
            expect(user.id).toMatch(/^usr_/);
        });

        it('should have valid email format', async () => {
            const user = await client.users.get(mockUser.id);
            expect(user.email).toMatch(/^[^\s@]+@[^\s@]+\.[^\s@]+$/);
        });
    });

    describe('Project contract', () => {
        it('should have required project fields', async () => {
            const project = await client.projects.get(mockProject.id);

            // Required fields
            expect(project).toHaveProperty('id');
            expect(project).toHaveProperty('name');
            expect(project).toHaveProperty('created_at');
            expect(project).toHaveProperty('updated_at');

            // Type checks
            expect(typeof project.id).toBe('string');
            expect(typeof project.name).toBe('string');
        });

        it('should have correct ID prefix', async () => {
            const project = await client.projects.get(mockProject.id);
            expect(project.id).toMatch(/^prj_/);
        });

        it('should have optional description field', async () => {
            const project = await client.projects.get(mockProject.id);

            if (project.description !== undefined) {
                expect(typeof project.description).toBe('string');
            }
        });
    });

    describe('Bucket contract', () => {
        it('should have required bucket fields', async () => {
            const bucket = await client.storage.buckets.get(mockBucket.id);

            // Required fields
            expect(bucket).toHaveProperty('id');
            expect(bucket).toHaveProperty('name');

            // Type checks
            expect(typeof bucket.id).toBe('string');
            expect(typeof bucket.name).toBe('string');
        });

        it('should have correct ID prefix', async () => {
            const bucket = await client.storage.buckets.get(mockBucket.id);
            expect(bucket.id).toMatch(/^bkt_/);
        });
    });

    describe('List response contract', () => {
        it('should have correct list response structure', async () => {
            const result = await client.users.list();

            expect(result).toHaveProperty('data');
            expect(result).toHaveProperty('has_more');
            expect(Array.isArray(result.data)).toBe(true);
            expect(typeof result.has_more).toBe('boolean');
        });

        it('should have nextCursor when hasMore is true', async () => {
            server.use(
                http.get(`${BASE_URL}/v1/users`, () => {
                    return HttpResponse.json({
                        data: [mockUser],
                        has_more: true,
                        next_cursor: 'cursor_123',
                    });
                })
            );

            const result = await client.users.list();

            expect(result.has_more).toBe(true);
            expect(result.next_cursor).toBeDefined();
            expect(typeof result.next_cursor).toBe('string');
        });
    });

    describe('Error response contract', () => {
        it('should have correct error response structure', async () => {
            server.use(
                http.get(`${BASE_URL}/v1/users/:id`, () => {
                    return HttpResponse.json({
                        error: {
                            message: 'User not found',
                            code: 'NOT_FOUND',
                            status: 404,
                        },
                    }, { status: 404 });
                })
            );

            try {
                await client.users.get('usr_nonexistent');
                expect.fail('Should have thrown');
            } catch (error) {
                expect(error).toBeInstanceOf(Error);
                expect((error as Error).message).toBeDefined();
            }
        });

        it('should include validation details for 400 errors', async () => {
            server.use(
                http.post(`${BASE_URL}/v1/users`, () => {
                    return HttpResponse.json({
                        error: {
                            message: 'Validation failed',
                            code: 'VALIDATION_ERROR',
                            status: 400,
                            details: [
                                { field: 'email', message: 'Invalid email format' },
                            ],
                        },
                    }, { status: 400 });
                })
            );

            try {
                await client.users.create({ name: 'Test', email: 'invalid' });
                expect.fail('Should have thrown');
            } catch (error) {
                expect(error).toBeInstanceOf(Error);
            }
        });
    });

    describe('Timestamp formats', () => {
        it('should return dates as Date objects or ISO strings', async () => {
            const user = await client.users.get(mockUser.id);

            // The SDK should parse dates - check if it's a Date or valid ISO string
            const createdAt = user.created_at;

            if (typeof createdAt === 'string') {
                expect(new Date(createdAt).getTime()).not.toBeNaN();
            } else {
                // Should be string
                expect(typeof createdAt).toBe('string');
            }
        });
    });
});

File: packages/go/luna/contracts/api_contract_test.go
package contracts_test

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"regexp"
	"testing"

	"github.com/eclipse-softworks/luna-sdk-go/luna"
	"github.com/eclipse-softworks/luna-sdk-go/luna/testutil"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestUserContract(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(testutil.MockUser)
	}))
	defer server.Close()

	client, err := luna.NewClient(
		luna.WithAPIKey("lk_test_12345678901234567890123456789012"),
		luna.WithBaseURL(server.URL),
	)
	require.NoError(t, err)

	t.Run("user has required fields", func(t *testing.T) {
		user, err := client.Users().Get(context.Background(), "usr_123456789")

		require.NoError(t, err)
		assert.NotEmpty(t, user.ID)
		assert.NotEmpty(t, user.Name)
		assert.NotEmpty(t, user.Email)
		assert.NotNil(t, user.CreatedAt)
		assert.NotNil(t, user.UpdatedAt)
	})

	t.Run("user ID has correct prefix", func(t *testing.T) {
		user, err := client.Users().Get(context.Background(), "usr_123456789")

		require.NoError(t, err)
		assert.Regexp(t, regexp.MustCompile(`^usr_`), user.ID)
	})

	t.Run("user email is valid format", func(t *testing.T) {
		user, err := client.Users().Get(context.Background(), "usr_123456789")

		require.NoError(t, err)
		emailRegex := regexp.MustCompile(`^[^\s@]+@[^\s@]+\.[^\s@]+$`)
		assert.Regexp(t, emailRegex, user.Email)
	})
}

func TestProjectContract(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(testutil.MockProject)
	}))
	defer server.Close()

	client, err := luna.NewClient(
		luna.WithAPIKey("lk_test_12345678901234567890123456789012"),
		luna.WithBaseURL(server.URL),
	)
	require.NoError(t, err)

	t.Run("project has required fields", func(t *testing.T) {
		project, err := client.Projects().Get(context.Background(), "prj_123456789")

		require.NoError(t, err)
		assert.NotEmpty(t, project.ID)
		assert.NotEmpty(t, project.Name)
		assert.NotNil(t, project.CreatedAt)
		assert.NotNil(t, project.UpdatedAt)
	})

	t.Run("project ID has correct prefix", func(t *testing.T) {
		project, err := client.Projects().Get(context.Background(), "prj_123456789")

		require.NoError(t, err)
		assert.Regexp(t, regexp.MustCompile(`^prj_`), project.ID)
	})

	t.Run("project description is optional", func(t *testing.T) {
		// Create server that returns project without description
		noDescServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			project := testutil.MockProject
			project["description"] = nil
			json.NewEncoder(w).Encode(project)
		}))
		defer noDescServer.Close()

		noDescClient, err := luna.NewClient(
			luna.WithAPIKey("lk_test_12345678901234567890123456789012"),
			luna.WithBaseURL(noDescServer.URL),
		)
		require.NoError(t, err)

		project, err := noDescClient.Projects().Get(context.Background(), "prj_123456789")

		require.NoError(t, err)
		assert.NotEmpty(t, project.ID)
		// Description can be nil or empty
	})
}

func TestBucketContract(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(testutil.MockListResponse(testutil.MockBuckets, false, ""))
	}))
	defer server.Close()

	client, err := luna.NewClient(
		luna.WithAPIKey("lk_test_12345678901234567890123456789012"),
		luna.WithBaseURL(server.URL),
	)
	require.NoError(t, err)

	t.Run("bucket list returns buckets", func(t *testing.T) {
		buckets, err := client.Storage().Buckets.List(context.Background())

		require.NoError(t, err)
		assert.NotNil(t, buckets)
		assert.Greater(t, len(buckets.Data), 0)
	})

	t.Run("bucket has required fields", func(t *testing.T) {
		buckets, err := client.Storage().Buckets.List(context.Background())

		require.NoError(t, err)
		if len(buckets.Data) > 0 {
			bucket := buckets.Data[0]
			assert.NotEmpty(t, bucket.ID)
			assert.NotEmpty(t, bucket.Name)
			assert.Regexp(t, regexp.MustCompile(`^bkt_`), bucket.ID)
		}
	})
}

func TestListResponseContract(t *testing.T) {
	t.Run("list response has correct structure", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(testutil.MockListResponse(testutil.MockUsers, false, ""))
		}))
		defer server.Close()

		client, err := luna.NewClient(
			luna.WithAPIKey("lk_test_12345678901234567890123456789012"),
			luna.WithBaseURL(server.URL),
		)
		require.NoError(t, err)

		result, err := client.Users().List(context.Background(), nil)

		require.NoError(t, err)
		assert.NotNil(t, result.Data)
		// HasMore should be a boolean (checked by type system)
	})

	t.Run("list response includes next_cursor when has_more is true", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(testutil.MockListResponse(testutil.MockUsers, true, "cursor_abc123"))
		}))
		defer server.Close()

		client, err := luna.NewClient(
			luna.WithAPIKey("lk_test_12345678901234567890123456789012"),
			luna.WithBaseURL(server.URL),
		)
		require.NoError(t, err)

		result, err := client.Users().List(context.Background(), nil)

		require.NoError(t, err)
		assert.True(t, result.HasMore)
		assert.NotEmpty(t, result.NextCursor)
	})
}

func TestErrorResponseContract(t *testing.T) {
	t.Run("error response has correct structure", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusNotFound)
			json.NewEncoder(w).Encode(testutil.MockErrorNotFound)
		}))
		defer server.Close()

		client, err := luna.NewClient(
			luna.WithAPIKey("lk_test_12345678901234567890123456789012"),
			luna.WithBaseURL(server.URL),
		)
		require.NoError(t, err)

		_, err = client.Users().Get(context.Background(), "usr_nonexistent")

		require.Error(t, err)
		// Error should have message
		assert.NotEmpty(t, err.Error())
	})

	t.Run("validation error includes details", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusBadRequest)
			json.NewEncoder(w).Encode(testutil.MockErrorValidation)
		}))
		defer server.Close()

		client, err := luna.NewClient(
			luna.WithAPIKey("lk_test_12345678901234567890123456789012"),
			luna.WithBaseURL(server.URL),
		)
		require.NoError(t, err)

		_, err = client.Users().Create(context.Background(), luna.UserCreate{
			Name:  "Test User",
			Email: "test@example.com",
		})

		require.Error(t, err)
		// Error message should contain validation info
		assert.Contains(t, err.Error(), "Validation")
	})
}

func TestTimestampFormats(t *testing.T) {
	t.Run("timestamps are valid", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(testutil.MockUser)
		}))
		defer server.Close()

		client, err := luna.NewClient(
			luna.WithAPIKey("lk_test_12345678901234567890123456789012"),
			luna.WithBaseURL(server.URL),
		)
		require.NoError(t, err)

		user, err := client.Users().Get(context.Background(), "usr_123456789")

		require.NoError(t, err)
		// CreatedAt should be a valid time
		assert.False(t, user.CreatedAt.IsZero())
	})
}
